nomos-libp2p/src/behaviour/nat/inner.rs

```diff
@@ -0,0 +1,211 @@
+                connection_id,
+                maybe_peer,
+                addresses,
+                effective_role,
+            )
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        role_override: Endpoint,
+        port_use: PortUse,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.autonat_client_behaviour
+            .handle_established_outbound_connection(
+                connection_id,
+                peer,
+                addr,
+                role_override,
+                port_use,
+            )
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        self.state_machine.on_event(event);
+        self.autonat_client_behaviour.on_swarm_event(event);
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: THandlerOutEvent<Self>,
+    ) {
+        self.autonat_client_behaviour
+            .on_connection_handler_event(peer_id, connection_id, event);
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
+        if let Poll::Ready(to_swarm) = self.autonat_client_behaviour.poll(cx) {
+            if let ToSwarm::GenerateEvent(event) = &to_swarm {
+                self.state_machine.on_event(event);
+            }
+
+            return Poll::Ready(to_swarm.map_out(Either::Left));
+        }
+
+        if let Poll::Ready(to_swarm) = self.address_mapper_behaviour.poll(cx) {
+            if let ToSwarm::GenerateEvent(event) = &to_swarm {
+                self.state_machine.on_event(event);
+            }
+
+            return Poll::Ready(to_swarm.map_out(Either::Right).map_in(Either::Right));
+        }
+
+        if let Poll::Ready(Some(_addr)) = self.next_autonat_client_tick.poll_unpin(cx) {
+            // TODO: This is a placeholder for the missing API of the
+            // autonat client
+            // self.autonat_client_behaviour.retest_address(addr);
+        }
+
+        if let Poll::Ready(Some(command)) = self.command_rx.poll_recv(cx) {
+            match command {
+                Command::ScheduleAutonatClientTest(addr) => {
+                    self.next_autonat_client_tick = OptionFuture::from(Some(
+                        tokio::time::sleep(self.autonat_client_tick_interval)
+                            .map(|()| addr)
+                            .boxed(),
+                    ))
+                    .fuse();
+                }
+                Command::MapAddress(addr) => {
+                    self.address_mapper_behaviour.try_map_address(addr);
+                }
+                Command::NewExternalAddrCandidate(addr) => {
+                    return Poll::Ready(ToSwarm::NewExternalAddrCandidate(addr));
+                }
+            }
+        }
+
+        Poll::Pending
```

Comment:
Do we need to awake the waker here or in some of the branches? What happens if we return `Pending` but do not awake the waker in case, for example, we poll a command from `self.command_rx`?